org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -15,137 +15,112 @@
 import java.util.Collections;
 import java.util.List;
 
+import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.NullProgressMonitor;
-import org.eclipse.jdt.core.IBuffer;
 import org.eclipse.jdt.core.ICompilationUnit;
-import org.eclipse.jdt.core.JavaModelException;
-import org.eclipse.jdt.core.compiler.IProblem;
 import org.eclipse.jdt.core.dom.CompilationUnit;
 import org.eclipse.jdt.ls.core.internal.JDTUtils;
 import org.eclipse.jdt.ls.core.internal.SharedASTProvider;
 import org.eclipse.jdt.ls.core.internal.TextEditConverter;
 import org.eclipse.jdt.ls.core.internal.corrections.DiagnosticsHelper;
-import org.eclipse.jdt.ls.core.internal.corrections.UnusedCodeCorrections;
+import org.eclipse.jdt.ls.core.internal.corrections.IProblemLocation;
+import org.eclipse.jdt.ls.core.internal.corrections.InnovationContext;
+import org.eclipse.jdt.ls.core.internal.corrections.ProblemLocation;
+import org.eclipse.jdt.ls.core.internal.corrections.QuickFixProcessor;
+import org.eclipse.jdt.ls.core.internal.corrections.proposals.CUCorrectionProposal;
 import org.eclipse.lsp4j.CodeActionParams;
 import org.eclipse.lsp4j.Command;
 import org.eclipse.lsp4j.Diagnostic;
-import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4j.DiagnosticSeverity;
 import org.eclipse.lsp4j.WorkspaceEdit;
+import org.eclipse.ltk.core.refactoring.TextChange;
+import org.eclipse.text.edits.TextEdit;
 
-/**
- * @author Gorkem Ercan
- *
- */
 public class CodeActionHandler {
 
 	/**
 	 *
 	 */
 	public static final String COMMAND_ID_APPLY_EDIT = "java.apply.workspaceEdit";
 
-
+	private QuickFixProcessor quickFixProcessor = new QuickFixProcessor();
 
 	/**
 	 * @param params
 	 * @return
 	 */
 	public List<Command> getCodeActionCommands(CodeActionParams params) {
```

Comment:
add IProgressMonitor parameter.

In JDTLanguageServer, call
```
return computeAsync((cc) -> handler. getCodeActionCommands(params, toMonitor(cc)));
```

so you can handle cancellations